packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol

```diff
@@ -474,15 +475,39 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
         tokenCurrency.burn(numTokens.rawValue);
     }
 
+    /**
+     * @notice Burns `numTokens` of `tokenCurrency` to decrease sponsors position size, without sending back any
+     * `collateralCurrency`. This is done by a sponsor to increase position CR.
+     * @dev Can only be called by token sponsor. This contract must be approved to spend `numTokens` of `tokenCurrency`.
+     * @param numTokens is the number of tokens to be burnt for a commensurate amount of collateral.
+     */
     function repay(FixedPoint.Unsigned memory numTokens)
         public
         notEmergencyShutdown()
         noPendingWithdrawal(msg.sender)
         fees()
         nonReentrant()
-        returns (FixedPoint.Unsigned memory amountRepaired)
     {
-        // TODO: implementation
+        PositionData storage positionData = _getPositionData(msg.sender);
+        require(!numTokens.isGreaterThan(positionData.tokensOutstanding), "Invalid token amount");
+
+        // Decrease the sponsors position tokens size. Ensure it is above the min sponsor size OR all tokens are repaid.
+        FixedPoint.Unsigned memory newTokenCount = positionData.tokensOutstanding.sub(numTokens);
+        require(
+            newTokenCount.isGreaterThanOrEqual(minSponsorTokens) ||
+                newTokenCount.isEqual(FixedPoint.fromUnscaledUint(0)),
+            "Below minimum sponsor position"
+        );
+        positionData.tokensOutstanding = newTokenCount;
+
+        // Update the totalTokensOutstanding after redemption.
+        totalTokensOutstanding = totalTokensOutstanding.sub(numTokens);
```

Comment:
Not really something we have to consider for this PR, but it might be nice to think about refactoring the internals (or the interface too!) to something like what @daywiss suggested where the user just tells us how they want to change their position (+- tokens, +- collateral). This method makes me think that something like that could make sense since this is another special case of that general pattern. Not actionable, just something to think on.

Reply:
That's a good idea as long as you factored GCR and CR's in. Not sure if those generalize well but I think they would